File: frontend/src/types/media.ts
export type ProcessingStatus = 'pending' | 'processing' | 'ready' | 'failed';
export type MediaType = 'image' | 'video';

export interface MediaAsset {
  id: string;
  project_id: string;
  filename: string;
  original_filename: string;
  media_type: MediaType;
  processing_status: ProcessingStatus;
  processing_error?: string;
  width?: number;
  height?: number;
  duration_ms?: number;
  fps?: number;
  file_size: number;
  thumbnail_url?: string;
  sort_order: number;
  created_at: string;
  processed_at?: string;
}

export interface AudioTrack {
  id: string;
  filename: string;
  duration_ms: number;
  sample_rate?: number;
  bpm?: number;
  analysis_status: 'queued' | 'processing' | 'complete' | 'failed';
  analysis_error?: string;
}

export interface MediaUploadResponse {
  uploaded: MediaAsset[];
  failed: { filename: string; error: string }[];
  total_uploaded: number;
}

export interface BeatsStatus {
  has_audio: boolean;
  analysis_status?: 'queued' | 'processing' | 'complete' | 'failed';
  analysis_error?: string;
  beat_count?: number;
  bpm?: number;
}

export interface Project {
  id: string;
  name: string;
  description?: string;
  media_assets: MediaAsset[];
  audio_track?: AudioTrack | null;
  created_at: string;
  updated_at: string;
}

File: backend/app/schemas/media.py
"""
Pydantic schemas for Media endpoints.

Includes request/response models for media asset operations.
"""

from datetime import datetime
from typing import List, Literal, Optional

from pydantic import BaseModel, Field


# --- Processing Status Type ---

ProcessingStatus = Literal["pending", "processing", "ready", "failed"]
MediaType = Literal["image", "video"]


# --- Response Schemas ---


class MediaUploadItem(BaseModel):
    """Single uploaded media asset response."""

    id: str
    filename: str
    media_type: MediaType
    processing_status: ProcessingStatus
    width: Optional[int] = None
    height: Optional[int] = None
    duration_ms: Optional[int] = None
    fps: Optional[float] = None
    file_size: int

    class Config:
        from_attributes = True


class MediaUploadResponse(BaseModel):
    """Response for media upload endpoint."""

    uploaded: List[MediaUploadItem]
    failed: List[dict] = Field(
        default_factory=list,
        description="List of files that failed to upload with error details"
    )
    total_uploaded: int


class MediaAssetResponse(BaseModel):
    """Full media asset details response."""

    id: str
    project_id: str
    filename: str
    media_type: MediaType
    processing_status: ProcessingStatus
    width: Optional[int] = None
    height: Optional[int] = None
    duration_ms: Optional[int] = None
    fps: Optional[float] = None
    file_size: int
    thumbnail_url: Optional[str] = None
    sort_order: int
    created_at: datetime
    processed_at: Optional[datetime] = None
    processing_error: Optional[str] = None

    class Config:
        from_attributes = True


# --- Request Schemas ---


class MediaReorderRequest(BaseModel):
    """Request to reorder media assets."""

    order: List[str] = Field(
        ...,
        min_length=1,
        description="List of media asset IDs in the new order"
    )


class MediaReorderItem(BaseModel):
    """Single item in reorder response."""

    id: str
    sort_order: int


class MediaReorderResponse(BaseModel):
    """Response for media reorder endpoint."""

    success: bool
    new_order: List[MediaReorderItem]
    timeline_invalidated: bool
